### DISCUSSION TITLE
Unable to Resolve SSH Invalid Credentials Warning in BTCPay Server

### DISCUSSION THREAD

**Discussion by ealvar13** [btcpayserver/btcpayserver#5333] [Technical Support]
Hi Everyone : )

I’ve been investigating the “Maintenance feature requires access to SSH properly configured in BTCPay Server configuration.” warning message.

I have SSH setup on my server.  It is running Ubuntu 22.04 and a dockerized version of BTCPay version v1.11.2+1c5fcfe09

I followed the docker installation guide, including setting the environment variable:
export BTCPAY_ENABLE_SSH=true

This is the guide I followed
https://github.com/btcpayserver/btcpayserver-docker#full-installation-for-technical-users 

Investigating the logs, I get this message repeatedly:
023-09-15 21:08:28.597 +00:00 [INF] SSH settings detected, testing connection to root@host.docker.internal on port 22 ...
2023-09-15 21:08:28.706 +00:00 [WRN] SSH invalid credentials (Permission denied (publickey).)
2023-09-15 21:08:28.706 +00:00 [WRN] Retrying SSH connection in 600 seconds

I’ve tried a number of potential fixes for this:
https://docs.btcpayserver.org/FAQ/ServerSettings/#error-btcpay-sshkeyfile-is-not-set-when-running-the-docker-install-or-unable-to-update-through-server-settings-maintenance
https://github.com/btcpayserver/btcpayserver-docker/issues/198 
https://github.com/btcpayserver/btcpayserver/issues/3833 

Do you all have any suggestions on what to try next?  I’m out of ideas.

Thanks so much in advance!

The following is a lot of additional info in case it is helpful:
----------------------------------------------------------------------------------------

I have a UFW running but it allows traffic on:
22/tcp                     ALLOW       Anywhere
22/tcp (v6)                ALLOW       Anywhere (v6)

Here are my current environment variables:
BTCPAY_ADDITIONAL_HOSTS=
BTCPAY_ENV_FILE=/root/BTCPayServer/.env
BTCPAY_SSHAUTHORIZEDKEYS=/datadir/host_authorized_keys
BTCPAY_SSHKEYFILE=/datadir/host_id_ed25519
BTCPAY_SSHTRUSTEDFINGERPRINTS=
BTCPAY_DOCKER_COMPOSE=/root/BTCPayServer/btcpayserver-docker/Generated/docker-compose.generated.yml
BTCPAY_HOST=testnet-btcpay.velascommerce.com
BTCPAY_ANNOUNCEABLE_HOST=testnet-btcpay.velascommerce.com
BTCPAY_PROTOCOL=https
BTCPAY_ENABLE_SSH=true
BTCPAYGEN_EXCLUDE_FRAGMENTS=
BTCPAYGEN_LIGHTNING=lnd
BTCPAY_BASE_DIRECTORY=/root/BTCPayServer
BTCPAY_HOST_SSHAUTHORIZEDKEYS=/root/.ssh/authorized_keys
BTCPAY_HOST_SSHKEYFILE=
BTCPAYGEN_ADDITIONAL_FRAGMENTS=opt-save-storage-xs;opt-lnd-config.custom;opt-lnd-config.custom;opt-lnd-config.custom;opt-lnd-config.custom;expose-lnd-grpc.custom;opt-lnd-config.custom
BTCPAY_IMAGE=
BTCPAY_CRYPTOS=btc
BTCPAYGEN_REVERSEPROXY=nginx
BTCPAYGEN_CRYPTO6=
BTCPAYGEN_CRYPTO7=
BTCPAYGEN_CRYPTO4=
BTCPAYGEN_CRYPTO5=
BTCPAYGEN_CRYPTO2=
BTCPAYGEN_CRYPTO3=
BTCPAYGEN_CRYPTO1=btc
BTCPAYGEN_CRYPTO8=
BTCPAYGEN_CRYPTO9=
BTCPAYGEN_OLD_PREGEN=false

Note the BTCPAY_HOST_SSHKEYFILE= environment variable is being ignored according the the setup script when I run the scripts and commands from the guidance here:
https://docs.btcpayserver.org/FAQ/ServerSettings/#error-btcpay-sshkeyfile-is-not-set-when-running-the-docker-install-or-unable-to-update-through-server-settings-maintenance 

Here are the details of running these suggestions:
root@localhost:~/BTCPayServer/btcpayserver-docker# . ./btcpay-setup.sh -i
WARNING: BTCPAY_HOST_SSHKEYFILE is now deprecated, use instead BTCPAY_ENABLE_SSH=true and run again '. btcpay-setup.sh -i'

-------SETUP-----------
Parameters passed:
BTCPAY_PROTOCOL:https
BTCPAY_HOST:testnet-btcpay.velascommerce.com
BTCPAY_ADDITIONAL_HOSTS:
REVERSEPROXY_HTTP_PORT:80
REVERSEPROXY_HTTPS_PORT:443
REVERSEPROXY_DEFAULT_HOST:none
LIBREPATRON_HOST:
ZAMMAD_HOST:
WOOCOMMERCE_HOST:
BTCTRANSMUTER_HOST:
CHATWOOT_HOST:
BTCPAY_ENABLE_SSH:true
BTCPAY_HOST_SSHKEYFILE:/root/.ssh/id_rsa_btcpay
LETSENCRYPT_EMAIL:
NBITCOIN_NETWORK:testnet
LIGHTNING_ALIAS:Velas-Test-BTCPay-Server
BTCPAYGEN_CRYPTO1:btc
BTCPAYGEN_CRYPTO2:
BTCPAYGEN_CRYPTO3:
BTCPAYGEN_CRYPTO4:
BTCPAYGEN_CRYPTO5:
BTCPAYGEN_CRYPTO6:
BTCPAYGEN_CRYPTO7:
BTCPAYGEN_CRYPTO8:
BTCPAYGEN_CRYPTO9:
BTCPAYGEN_REVERSEPROXY:nginx
BTCPAYGEN_LIGHTNING:lnd
BTCPAYGEN_ADDITIONAL_FRAGMENTS:opt-save-storage-xs;opt-lnd-config.custom;opt-lnd-config.custom;opt-lnd-config.custom;opt-lnd-config.custom;expose-lnd-grpc.custom;opt-lnd-config.custom
BTCPAYGEN_EXCLUDE_FRAGMENTS:
BTCPAY_IMAGE:
ACME_CA_URI:production
TOR_RELAY_NICKNAME:
TOR_RELAY_EMAIL:
PIHOLE_SERVERIP:
FIREFLY_HOST:
----------------------
Additional exported variables:
BTCPAY_DOCKER_COMPOSE=/root/BTCPayServer/btcpayserver-docker/Generated/docker-compose.generated.yml
BTCPAY_BASE_DIRECTORY=/root/BTCPayServer
BTCPAY_ENV_FILE=/root/BTCPayServer/.env
BTCPAYGEN_OLD_PREGEN=false
BTCPAY_SSHKEYFILE=/datadir/host_id_ed25519
BTCPAY_SSHAUTHORIZEDKEYS=/datadir/host_authorized_keys
BTCPAY_HOST_SSHAUTHORIZEDKEYS:/root/.ssh/authorized_keys
BTCPAY_SSHTRUSTEDFINGERPRINTS:
BTCPAY_CRYPTOS:btc
BTCPAY_ANNOUNCEABLE_HOST:testnet-btcpay.velascommerce.com
----------------------

BTCPay Server environment variables successfully saved in /etc/profile.d/btcpay-env.sh

BTCPay Server docker-compose parameters saved in /root/BTCPayServer/.env

Using default tag: latest
latest: Pulling from btcpayserver/docker-compose-generator
Digest: sha256:3a66d72869c71e032396bd4e0cf997533c2118467be1a4706dcd2f11beb3ac8e
Status: Image is up to date for btcpayserver/docker-compose-generator:latest
docker.io/btcpayserver/docker-compose-generator:latest
Crypto: btc
Lightning: lnd
ReverseProxy: nginx
Generating /app/Generated/docker-compose.generated.yml
Selected fragments:
        nginx
        btcpayserver
        bitcoin
        bitcoin-lnd
        opt-save-storage-xs
        opt-lnd-config.custom
        expose-lnd-grpc.custom
        opt-mempoolfullrbf
        opt-add-tor
        nginx-https
        opt-add-zmq
        nbxplorer
        postgres
        btcpayserver-nginx
Generated /app/Generated/pull-images.sh
Generated /app/Generated/save-images.sh
Generated /app/Generated/docker-compose.generated.yml

Adding btcpayserver.service to systemd
BTCPay Server systemd configured in /etc/systemd/system/btcpayserver.service

BTCPay Server starting... this can take 5 to 10 minutes...
BTCPay Server started
tor is up-to-date
generated_postgres_1 is up-to-date
btcpayserver_bitcoind is up-to-date
tor-gen is up-to-date
generated_nbxplorer_1 is up-to-date
btcpayserver_lnd_bitcoin is up-to-date
nginx is up-to-date
generated_bitcoin_rtl_1 is up-to-date
generated_btcpayserver_1 is up-to-date
nginx-gen is up-to-date
letsencrypt-nginx-proxy-companion is up-to-date
WARNING: BTCPAY_HOST_SSHKEYFILE is now deprecated, use instead BTCPAY_ENABLE_SSH=true and run again '. btcpay-setup.sh -i'
Copying /datadir/host_id_ed25519 to BTCPayServer container
Successfully copied 4.1kB to 2d7d14287fa1:/datadir/host_id_ed25519
Installed bitcoin-cli.sh to /usr/local/bin: Command line for your Bitcoin instance
Installed bitcoin-lncli.sh to /usr/local/bin: Command line for your Bitcoin LND instance
Installed btcpay-clean.sh to /usr/local/bin: Command line for deleting old unused docker images
Installed btcpay-down.sh to /usr/local/bin: Command line for stopping all services related to BTCPay Server
Installed btcpay-restart.sh to /usr/local/bin: Command line for restarting all services related to BTCPay Server
Installed btcpay-setup.sh to /usr/local/bin: Command line for restarting all services related to BTCPay Server
Installed btcpay-up.sh to /usr/local/bin: Command line for starting all services related to BTCPay Server
Installed btcpay-admin.sh to /usr/local/bin: Command line for some administrative operation in BTCPay Server
Installed btcpay-update.sh to /usr/local/bin: Command line for updating your BTCPay Server to the latest commit of this repository
Installed changedomain.sh to /usr/local/bin: Command line for changing the external domain of your BTCPay Server

**Comment by dennisreimann** [comment]
Can you check if this helps?

```bash
rm /var/lib/docker/volumes/generated_btcpay_datadir/_data/host_*
. btcpay-setup.sh -i
```

**Reply by ealvar13** [reply]
Hi Dennis!

Thanks for the suggestion!  Unfortunately it doesn’t seem to be working for me.

I tried two ways.  I ran:
$ rm /var/lib/docker/volumes/generated_btcpay_datadir/_data/host_*
$ . btcpay-setup.sh -i

And 
$ rm /var/lib/docker/volumes/generated_btcpay_datadir/_data/host_*
$ . ./btcpay-setup.sh -i

The results were similar.  I’m still seeing this in the log files:
2023-09-21 15:26:42.501 +00:00 [INF] SSH settings detected, testing connection to root@host.docker.internal on port 22 ...
2023-09-21 15:26:42.629 +00:00 [WRN] SSH invalid credentials (Permission denied (publickey).)
2023-09-21 15:26:42.629 +00:00 [WRN] Retrying SSH connection in 600 seconds
2023-09-21 15:36:42.628 +00:00 [INF] SSH settings detected, testing connection to root@host.docker.internal on port 22 …

Any ideas on something else I can try?  Should I attempt to remove the BTCPAY_HOST_SSHKEYFILE env variable?

Here is the detailed output from running the commands:
root@localhost:~/BTCPayServer/btcpayserver-docker# . ./btcpay-setup.sh -i
WARNING: BTCPAY_HOST_SSHKEYFILE is now deprecated, use instead BTCPAY_ENABLE_SSH=true and run again '. btcpay-setup.sh -i'

-------SETUP-----------
Parameters passed:
BTCPAY_PROTOCOL:https
BTCPAY_HOST:testnet-btcpay.velascommerce.com
BTCPAY_ADDITIONAL_HOSTS:
REVERSEPROXY_HTTP_PORT:80
REVERSEPROXY_HTTPS_PORT:443
REVERSEPROXY_DEFAULT_HOST:none
LIBREPATRON_HOST:
ZAMMAD_HOST:
WOOCOMMERCE_HOST:
BTCTRANSMUTER_HOST:
CHATWOOT_HOST:
BTCPAY_ENABLE_SSH:true
BTCPAY_HOST_SSHKEYFILE:/root/.ssh/id_rsa_btcpay
LETSENCRYPT_EMAIL:
NBITCOIN_NETWORK:testnet
LIGHTNING_ALIAS:Velas-Test-BTCPay-Server
BTCPAYGEN_CRYPTO1:btc
BTCPAYGEN_CRYPTO2:
BTCPAYGEN_CRYPTO3:
BTCPAYGEN_CRYPTO4:
BTCPAYGEN_CRYPTO5:
BTCPAYGEN_CRYPTO6:
BTCPAYGEN_CRYPTO7:
BTCPAYGEN_CRYPTO8:
BTCPAYGEN_CRYPTO9:
BTCPAYGEN_REVERSEPROXY:nginx
BTCPAYGEN_LIGHTNING:lnd
BTCPAYGEN_ADDITIONAL_FRAGMENTS:opt-save-storage-xs;opt-lnd-config.custom;opt-lnd-config.custom;opt-lnd-config.custom;opt-lnd-config.custom;expose-lnd-grpc.custom;opt-lnd-config.custom
BTCPAYGEN_EXCLUDE_FRAGMENTS:
BTCPAY_IMAGE:
ACME_CA_URI:production
TOR_RELAY_NICKNAME:
TOR_RELAY_EMAIL:
PIHOLE_SERVERIP:
FIREFLY_HOST:
----------------------
Additional exported variables:
BTCPAY_DOCKER_COMPOSE=/root/BTCPayServer/btcpayserver-docker/Generated/docker-compose.generated.yml
BTCPAY_BASE_DIRECTORY=/root/BTCPayServer
BTCPAY_ENV_FILE=/root/BTCPayServer/.env
BTCPAYGEN_OLD_PREGEN=false
BTCPAY_SSHKEYFILE=/datadir/host_id_ed25519
BTCPAY_SSHAUTHORIZEDKEYS=/datadir/host_authorized_keys
BTCPAY_HOST_SSHAUTHORIZEDKEYS:/root/.ssh/authorized_keys
BTCPAY_SSHTRUSTEDFINGERPRINTS:
BTCPAY_CRYPTOS:btc
BTCPAY_ANNOUNCEABLE_HOST:testnet-btcpay.velascommerce.com
----------------------

BTCPay Server environment variables successfully saved in /etc/profile.d/btcpay-env.sh

BTCPay Server docker-compose parameters saved in /root/BTCPayServer/.env

Using default tag: latest
latest: Pulling from btcpayserver/docker-compose-generator
Digest: sha256:3a66d72869c71e032396bd4e0cf997533c2118467be1a4706dcd2f11beb3ac8e
Status: Image is up to date for btcpayserver/docker-compose-generator:latest
docker.io/btcpayserver/docker-compose-generator:latest
Crypto: btc
Lightning: lnd
ReverseProxy: nginx
Generating /app/Generated/docker-compose.generated.yml
Selected fragments:
        nginx
        btcpayserver
        bitcoin
        bitcoin-lnd
        opt-save-storage-xs
        opt-lnd-config.custom
        expose-lnd-grpc.custom
        opt-mempoolfullrbf
        opt-add-tor
        nginx-https
        opt-add-zmq
        nbxplorer
        postgres
        btcpayserver-nginx
Generated /app/Generated/pull-images.sh
Generated /app/Generated/save-images.sh
Generated /app/Generated/docker-compose.generated.yml

Adding btcpayserver.service to systemd
BTCPay Server systemd configured in /etc/systemd/system/btcpayserver.service

BTCPay Server starting... this can take 5 to 10 minutes...
BTCPay Server started
tor is up-to-date
generated_postgres_1 is up-to-date
tor-gen is up-to-date
btcpayserver_bitcoind is up-to-date
generated_nbxplorer_1 is up-to-date
btcpayserver_lnd_bitcoin is up-to-date
nginx is up-to-date
generated_bitcoin_rtl_1 is up-to-date
generated_btcpayserver_1 is up-to-date
nginx-gen is up-to-date
letsencrypt-nginx-proxy-companion is up-to-date
WARNING: BTCPAY_HOST_SSHKEYFILE is now deprecated, use instead BTCPAY_ENABLE_SSH=true and run again '. btcpay-setup.sh -i'
Copying /datadir/host_id_ed25519 to BTCPayServer container
Successfully copied 4.1kB to 2d7d14287fa1:/datadir/host_id_ed25519
Installed bitcoin-cli.sh to /usr/local/bin: Command line for your Bitcoin instance
Installed bitcoin-lncli.sh to /usr/local/bin: Command line for your Bitcoin LND instance
Installed btcpay-clean.sh to /usr/local/bin: Command line for deleting old unused docker images
Installed btcpay-down.sh to /usr/local/bin: Command line for stopping all services related to BTCPay Server
Installed btcpay-restart.sh to /usr/local/bin: Command line for restarting all services related to BTCPay Server
Installed btcpay-setup.sh to /usr/local/bin: Command line for restarting all services related to BTCPay Server
Installed btcpay-up.sh to /usr/local/bin: Command line for starting all services related to BTCPay Server
Installed btcpay-admin.sh to /usr/local/bin: Command line for some administrative operation in BTCPay Server
Installed btcpay-update.sh to /usr/local/bin: Command line for updating your BTCPay Server to the latest commit of this repository
Installed changedomain.sh to /usr/local/bin: Command line for changing the external domain of your BTCPay Server

**Comment by dennisreimann** [comment]
Can you also try this:

```bash
unset BTCPAY_HOST_SSHKEYFILE
export BTCPAY_ENABLE_SSH=true
. btcpay-setup.sh -i
```

**Reply by ealvar13** [reply]
Thanks again I really appreciate the help.  Unfortunately, the docker container still seems to be having connection problems.

**I ran this:**
root@localhost:~# cd BTCPayServer/btcpayserver-docker/
root@localhost:~/BTCPayServer/btcpayserver-docker# unset BTCPAY_HOST_SSHKEYFILE
root@localhost:~/BTCPayServer/btcpayserver-docker# export BTCPAY_ENABLE_SSH=true
root@localhost:~/BTCPayServer/btcpayserver-docker# . btcpay-setup.sh -i

**I no longer get the warnings about deprecated SSH settings, however the logs are still showing a connection problem:**
2023-09-21 19:18:33.715 +00:00 [INF] User '145dd46f-6b5d-468f-9cc0-5779574bae9a' logged in.
2023-09-21 19:26:45.661 +00:00 [INF] SSH settings detected, testing connection to root@host.docker.internal on port 22 ...
2023-09-21 19:26:45.807 +00:00 [WRN] SSH invalid credentials (Permission denied (publickey).)
2023-09-21 19:26:45.807 +00:00 [WRN] Retrying SSH connection in 600 seconds

**Any other ideas?**

**Here’s the detailed input and output:**
root@localhost:~# cd BTCPayServer/btcpayserver-docker/
root@localhost:~/BTCPayServer/btcpayserver-docker# unset BTCPAY_HOST_SSHKEYFILE
root@localhost:~/BTCPayServer/btcpayserver-docker# export BTCPAY_ENABLE_SSH=true
root@localhost:~/BTCPayServer/btcpayserver-docker# . btcpay-setup.sh -i

-------SETUP-----------
Parameters passed:
BTCPAY_PROTOCOL:https
BTCPAY_HOST:testnet-btcpay.velascommerce.com
BTCPAY_ADDITIONAL_HOSTS:
REVERSEPROXY_HTTP_PORT:80
REVERSEPROXY_HTTPS_PORT:443
REVERSEPROXY_DEFAULT_HOST:none
LIBREPATRON_HOST:
ZAMMAD_HOST:
WOOCOMMERCE_HOST:
BTCTRANSMUTER_HOST:
CHATWOOT_HOST:
BTCPAY_ENABLE_SSH:true
BTCPAY_HOST_SSHKEYFILE:
LETSENCRYPT_EMAIL:
NBITCOIN_NETWORK:testnet
LIGHTNING_ALIAS:Velas-Test-BTCPay-Server
BTCPAYGEN_CRYPTO1:btc
BTCPAYGEN_CRYPTO2:
BTCPAYGEN_CRYPTO3:
BTCPAYGEN_CRYPTO4:
BTCPAYGEN_CRYPTO5:
BTCPAYGEN_CRYPTO6:
BTCPAYGEN_CRYPTO7:
BTCPAYGEN_CRYPTO8:
BTCPAYGEN_CRYPTO9:
BTCPAYGEN_REVERSEPROXY:nginx
BTCPAYGEN_LIGHTNING:lnd
BTCPAYGEN_ADDITIONAL_FRAGMENTS:opt-save-storage-xs;opt-lnd-config.custom;opt-lnd-config.custom;opt-lnd-config.custom;opt-lnd-config.custom;expose-lnd-grpc.custom;opt-lnd-config.custom
BTCPAYGEN_EXCLUDE_FRAGMENTS:
BTCPAY_IMAGE:
ACME_CA_URI:production
TOR_RELAY_NICKNAME:
TOR_RELAY_EMAIL:
PIHOLE_SERVERIP:
FIREFLY_HOST:
----------------------
Additional exported variables:
BTCPAY_DOCKER_COMPOSE=/root/BTCPayServer/btcpayserver-docker/Generated/docker-compose.generated.yml
BTCPAY_BASE_DIRECTORY=/root/BTCPayServer
BTCPAY_ENV_FILE=/root/BTCPayServer/.env
BTCPAYGEN_OLD_PREGEN=false
BTCPAY_SSHKEYFILE=/datadir/host_id_ed25519
BTCPAY_SSHAUTHORIZEDKEYS=/datadir/host_authorized_keys
BTCPAY_HOST_SSHAUTHORIZEDKEYS:/root/.ssh/authorized_keys
BTCPAY_SSHTRUSTEDFINGERPRINTS:
BTCPAY_CRYPTOS:btc
BTCPAY_ANNOUNCEABLE_HOST:testnet-btcpay.velascommerce.com
----------------------

BTCPay Server environment variables successfully saved in /etc/profile.d/btcpay-env.sh

BTCPay Server docker-compose parameters saved in /root/BTCPayServer/.env

Using default tag: latest
latest: Pulling from btcpayserver/docker-compose-generator
Digest: sha256:3a66d72869c71e032396bd4e0cf997533c2118467be1a4706dcd2f11beb3ac8e
Status: Image is up to date for btcpayserver/docker-compose-generator:latest
docker.io/btcpayserver/docker-compose-generator:latest
Crypto: btc
Lightning: lnd
ReverseProxy: nginx
Generating /app/Generated/docker-compose.generated.yml
Selected fragments:
        nginx
        btcpayserver
        bitcoin
        bitcoin-lnd
        opt-save-storage-xs
        opt-lnd-config.custom
        expose-lnd-grpc.custom
        opt-mempoolfullrbf
        opt-add-tor
        nginx-https
        opt-add-zmq
        nbxplorer
        postgres
        btcpayserver-nginx
Generated /app/Generated/pull-images.sh
Generated /app/Generated/save-images.sh
Generated /app/Generated/docker-compose.generated.yml

Adding btcpayserver.service to systemd
BTCPay Server systemd configured in /etc/systemd/system/btcpayserver.service

BTCPay Server starting... this can take 5 to 10 minutes...
BTCPay Server started
generated_postgres_1 is up-to-date
tor is up-to-date
generated_nbxplorer_1 is up-to-date
btcpayserver_bitcoind is up-to-date
tor-gen is up-to-date
btcpayserver_lnd_bitcoin is up-to-date
generated_btcpayserver_1 is up-to-date
generated_bitcoin_rtl_1 is up-to-date
nginx is up-to-date
nginx-gen is up-to-date
letsencrypt-nginx-proxy-companion is up-to-date
Installed bitcoin-cli.sh to /usr/local/bin: Command line for your Bitcoin instance
Installed bitcoin-lncli.sh to /usr/local/bin: Command line for your Bitcoin LND instance
Installed btcpay-clean.sh to /usr/local/bin: Command line for deleting old unused docker images
Installed btcpay-down.sh to /usr/local/bin: Command line for stopping all services related to BTCPay Server
Installed btcpay-restart.sh to /usr/local/bin: Command line for restarting all services related to BTCPay Server
Installed btcpay-setup.sh to /usr/local/bin: Command line for restarting all services related to BTCPay Server
Installed btcpay-up.sh to /usr/local/bin: Command line for starting all services related to BTCPay Server
Installed btcpay-admin.sh to /usr/local/bin: Command line for some administrative operation in BTCPay Server
Installed btcpay-update.sh to /usr/local/bin: Command line for updating your BTCPay Server to the latest commit of this repository
Installed changedomain.sh to /usr/local/bin: Command line for changing the external domain of your BTCPay Server

**Reply by dennisreimann** [reply]
Ok, this indicates a key mismatch. Can you check if the file `/root/.ssh/authorized_keys` contains the public key that's also in `host_id_ed25519.pub`?

```bash
cat /root/.ssh/authorized_keys
cat /var/lib/docker/volumes/generated_btcpay_datadir/_data/host_id_ed25519.pub 
```

If not, delete the existing `ssh-ed25519 KEY btcpayserver` line from `/root/.ssh/authorized_keys` and replace it with the line from the `host_id_ed25519.pub` file. That should make it work.

**Reply by ealvar13** [reply]
Thanks!  I think we’re getting somewhere.  When I checked the folder here: 
$ cat /var/lib/docker/volumes/generated_btcpay_datadir/_data/host_id_ed25519.pub

It was actually missing the pub key although the private key was there.  I regenerated the pub key and now the pub keys in the two directories match:
$ cat /root/.ssh/authorized_keys
$ cat /var/lib/docker/volumes/generated_btcpay_datadir/_data/host_id_ed25519.pub

Next I monitored the logs to see if this fixed the recurring error of SSH keys not found.  It retries every 600 seconds.

On review, the BTCPay Server detected the changes in SSH settings, however it still rejects the connection.  So it seems like we’re getting somewhere.

I tried running the setup script again but the result is the same.  Do you think I need to reboot the actual Linux server?

Any ideas on what I can try next?

Here’s an example of the new recurring error in the logs:
2023-09-22 13:11:00.206 +00:00 [INF] User '145dd46f-6b5d-468f-9cc0-5779574bae9a' logged in.
2023-09-22 13:16:58.561 +00:00 [INF] SSH settings detected, testing connection to root@host.docker.internal on port 22 ...
2023-09-22 13:16:58.720 +00:00 [WRN] SSH invalid credentials (Permission denied (publickey).)
2023-09-22 13:16:58.720 +00:00 [WRN] Retrying SSH connection in 600 seconds
2023-09-22 13:26:58.719 +00:00 [INF] SSH settings detected, testing connection to root@host.docker.internal on port 22 ...

**Reply by dennisreimann** [reply]
I assume it is either a permission problem (please check the file access and owners, [details here](https://docs.digitalocean.com/support/how-to-troubleshoot-ssh-authentication-issues/#fixing-key-permissions-and-ownership)) or the SSH agent might need a restart.

**Comment by ealvar13** [comment]
Hi Dennis!

Thanks for this suggestion.  I verified the permissions, added and added the SSH agent.  When I verified that this hadn’t solved the SSH authentication issue, I also completely rebooted the server to see if that would help.

So far, no luck, the BTCPay Server can’t authenticate using the SSH keys.  

**Any ideas what else I can try?**

Here’s all the details from following the above suggestions in case it helps diagnose the issue:

drwx------  2 root root 4.0K Sep 15 22:05 .ssh

Indicate that:
~/.ssh permissions should be 700
~/.ssh should be owned by your account

**Check.**

-rw------- 1 root root 4.7K Sep 20 12:00 authorized_keys

Indicates that:
~/.ssh/authorized_keys permissions should be 600
~/.ssh/authorized_keys should be owned by your account

**Check.**

Restarting the SSH agent:
$ eval $(ssh-agent -k)
root@localhost:~# eval $(ssh-agent -k)
SSH_AGENT_PID not set, cannot kill agent

**It looks like there was no ssh-agent running?**


$ ps aux | grep ssh-agent
root     1292121  0.0  0.1   6608  2152 pts/0    S+   21:38   0:00 grep --color=auto ssh-agent
**So there is no SSH agent running**

$ eval $(ssh-agent)
Agent pid 1292137

**Add the private key located here:**
/root/.ssh/id_rsa_btcpay
$ ssh-add /root/.ssh/id_rsa_btcpay
Identity added: /root/.ssh/id_rsa_btcpay (/root/.ssh/id_rsa_btcpay)

root@localhost:~/.ssh# ssh-add -l
3072 SHA256:miZGr1RHoslr/uZbSwpNSOiYVKVRd+X1awP8hgo1eyA /root/.ssh/id_rsa_btcpay (RSA)

**So it was added correctly.
Same problem per the logs:**
2023-09-22 21:37:04.938 +00:00 [WRN] Retrying SSH connection in 600 seconds
2023-09-22 21:47:04.938 +00:00 [INF] SSH settings detected, testing connection to root@host.docker.internal on port 22 ...
2023-09-22 21:47:05.070 +00:00 [WRN] SSH invalid credentials (Permission denied (publickey).)
2023-09-22 21:47:05.071 +00:00 [WRN] Retrying SSH connection in 600 seconds


**Rebooting the server to see if it helps.** 
Rebooting at 1749 UTC-4 20230922

**No luck.  BTCPay Server can’t find the SSH Pubkey**

2023-09-22 21:51:28.963 +00:00 [WRN] SSH invalid credentials (Permission denied (publickey).)
2023-09-22 21:51:28.963 +00:00 [WRN] Retrying SSH connection in 10 seconds
2023-09-22 21:51:29.428 +00:00 [INF] 1 pending payment requests being checked since last run
2023-09-22 21:51:29.964 +00:00 [INF] Starting
2023-09-22 21:51:38.961 +00:00 [INF] SSH settings detected, testing connection to root@host.docker.internal on port 22 ...
2023-09-22 21:51:39.163 +00:00 [WRN] SSH invalid credentials (Permission denied (publickey).)

**Comment by NicolasDorier** [comment]
@ealvar13 first of all: Don't worry if you can't make it works. The SSH access is only used for convenience to update through the UI. When you want to update you can just run `btcpay-update.sh` by yourself. Nothing of much importance is lost.

That said, I'd like to know what is going on.

From my working box, here is what I have:

### Private key permissions
```bash
docker exec generated_btcpayserver_1 bash -c 'ls -alh ${BTCPAY_SSHKEYFILE}*'
```
Output:
```
-rw------- 1 root root 399 Jul  8  2022 /datadir/host_id_ed25519
-rw-r--r-- 1 root root  94 Jul  8  2022 /datadir/host_id_ed25519.pub
```

### Public key content
```bash
docker exec generated_btcpayserver_1 bash -c 'cat $BTCPAY_SSHKEYFILE.pub'
```

```
ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIOYJEnUmwDW6FBtoN9CbPGRDxoGimfgzjyWZGEd0ZC+u btcpayserver
```

### Authorized Keys permissions

```bash
docker exec generated_btcpayserver_1 bash -c 'ls -alh $BTCPAY_SSHAUTHORIZEDKEYS'
```

```
-rw------- 1 root root 904 Jul  8  2022 /datadir/host_authorized_keys
```

### Authorized Keys content

```bash
docker exec generated_btcpayserver_1 bash -c 'cat $BTCPAY_SSHAUTHORIZEDKEYS'
```

```
...
ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIOYJEnUmwDW6FBtoN9CbPGRDxoGimfgzjyWZGEd0ZC+u btcpayserver
```

### Check that the authorized keys on the host is same as in container

```
cat ~/.ssh/authorized_keys
```
```
...
ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIOYJEnUmwDW6FBtoN9CbPGRDxoGimfgzjyWZGEd0ZC+u btcpayserver
```

### Troubleshoot
* Can you check that `cat ~/.ssh/authorized_keys` and `docker exec generated_btcpayserver_1 bash -c 'cat $BTCPAY_SSHAUTHORIZEDKEYS'` have same result?
* Can you check the permissions match?
* Can you check that `cat ~/.ssh/authorized_keys` has the same key as `docker exec generated_btcpayserver_1 bash -c 'cat $BTCPAY_SSHKEYFILE.pub'`?

**Comment by ealvar13** [comment]
Hi Nicolas,

Thanks for all your suggestions.  It seems like we’re so close to solving this but I just can’t quite understand what the issue is.

I tried all the suggestions to make sure the keys match and they have the right permissions.

Since it was using an RSA type key I thought that might be the source of the problem.  I generated an new ed25519 key pair, set the permissions, and added them to:
/root/.ssh/authorized_keys
and
/var/lib/docker/volumes/generated_btcpay_datadir/_data/btcpay_id_ed25519.pub

I also noticed that the host_authorized_keys file here was empty, so I added the new pub key to that:
/var/lib/docker/volumes/generated_btcpay_datadir/_data/host_authorized_keys

So now all these keys match and they have the correct permissions:
/root/.ssh/authorized_keys
/var/lib/docker/volumes/generated_btcpay_datadir/_data/btcpay_id_ed25519.pub
/var/lib/docker/volumes/generated_btcpay_datadir/_data/host_authorized_keys

I also confirmed that, from within the btcpay server docker container’s bash shell, I can connect to the local host computer using: 
root@b96e8d20c7a7:/datadir# ssh -i /datadir/btcpay_id_ed25519 -p 22 root@host.docker.internal
(this connection is successful)

When the error continued, I reran the setup script, regenerated the SSH agent, and restarted BTCPay server.

When I went back to check your advice, I ran: 
$ docker exec generated_btcpayserver_1 bash -c 'cat $BTCPAY_SSHKEYFILE.pub'

This showed that the BTCPAY_SSHKEYFILE.pub was still using the old RSA keys.  

So I double checked all my files and added the RSA keys to the host_authorized_keys file and I was able to SSH from docker into the local host using the RSA keys as well:

Do you all have any more ideas?  If you think it’s best to just leave this alone, since it doesn’t really affect anything other than having a warning message and clogging up the logs, that’s totally cool.

I appreciate all your help : )

**Reply by dennisreimann** [reply]
> I also noticed that the host_authorized_keys file here was empty, so I added the new pub key to that:
> /var/lib/docker/volumes/generated_btcpay_datadir/_data/host_authorized_keys

Did you restart BTCPay afterwards? By now this is the only thing I can think of.

**Reply by ealvar13** [reply]
Yep I tried that a bunch of times and also rebooted the actual server.  I can't think of anything else either.   Let's leave it for now.  This affects a mainnet BTCPay Server we run too.  But there really is no harm.   

Thanks for all your help.
